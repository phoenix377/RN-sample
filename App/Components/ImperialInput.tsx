import React, { useEffect } from 'react';
import {
  Text,
  TextInput,
  TextInputProps,
  useColorScheme,
  View,
} from 'react-native';
import { feetToMeters, metersToFeet } from '../utils';

import {
  CONTAINER,
  ERROR,
  LABEL,
  LABEL_CONTAINER,
  LABEL_DARK,
  TEXT_INPUT,
  TEXT_INPUT_DARK,
  TEXT_INPUT_ERROR,
} from './Input.styles';

type ImperialInputProps = TextInputProps & {
  label: string;
  error?: boolean;
};

const ImperialInput: React.FC<ImperialInputProps> = props => {
  const isDarkMode = useColorScheme() === 'dark';
  const [valueFt, setValueFt] = React.useState('');
  const [valueInch, setValueInch] = React.useState('');
  const onChangeText = props.onChangeText;
  useEffect(() => {
    const newValue: number = feetToMeters(
      (
        parseInt(valueFt || '0', 10) +
        parseInt(valueInch || '0', 10) / 12
      ).toString(),
    );
    console.log('props.onChangeText', newValue);
    onChangeText?.(newValue.toString());
  }, [valueFt, valueInch, onChangeText]);

  useEffect(() => {
    if (!props.value) {
      return;
    }
    console.log('props.value', props.value);
    const imp = metersToFeet(props.value || '0');
    const ft = Math.floor(imp);
    const inch = Math.round((imp - ft) * 12);
    setValueFt(ft.toString());
    setValueInch(inch.toString());
  }, [props.value]);

  return (
    <View style={[CONTAINER, props.error ? ERROR : null]}>
      <TextInput
        style={[
          isDarkMode ? TEXT_INPUT_DARK : TEXT_INPUT,
          props.error ? TEXT_INPUT_ERROR : null,
        ]}
        onBlur={props.onBlur}
        onChangeText={setValueFt}
        value={valueFt}
        placeholder={props.placeholder}
      />
      {props.label && (
        <View style={LABEL_CONTAINER}>
          <Text style={isDarkMode ? LABEL_DARK : LABEL}>ft</Text>
        </View>
      )}
      <TextInput
        style={[
          isDarkMode ? TEXT_INPUT_DARK : TEXT_INPUT,
          props.error ? TEXT_INPUT_ERROR : null,
        ]}
        onBlur={props.onBlur}
        onChangeText={setValueInch}
        value={valueInch}
        placeholder={props.placeholder}
      />
      {props.label && (
        <View style={LABEL_CONTAINER}>
          <Text style={isDarkMode ? LABEL_DARK : LABEL}>in</Text>
        </View>
      )}
    </View>
  );
};

export default ImperialInput;
