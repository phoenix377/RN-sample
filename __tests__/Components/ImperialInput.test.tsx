import 'react-native';

import React from 'react';
import renderer from 'react-test-renderer';

import ImperialInput from '../../App/Components/ImperialInput';

it('ImperialInput renders correctly', () => {
  const tree = renderer
    .create(
      <ImperialInput label="Height" value="180" onChangeText={jest.fn()} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
