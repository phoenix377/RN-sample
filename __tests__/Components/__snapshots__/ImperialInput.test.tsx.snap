// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImperialInput renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "#ccc",
        "flex": 1,
        "flexDirection": "row",
        "height": 40,
        "marginBottom": 10,
      },
      null,
    ]
  }
>
  <TextInput
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderColor": "gray",
          "borderRadius": 20,
          "borderWidth": 1,
          "flex": 1,
          "fontSize": 15,
          "height": 40,
          "paddingHorizontal": 20,
        },
        null,
      ]
    }
    value=""
  />
  <View
    style={
      Object {
        "width": 40,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 15,
          "marginLeft": 10,
        }
      }
    >
      ft
    </Text>
  </View>
  <TextInput
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderColor": "gray",
          "borderRadius": 20,
          "borderWidth": 1,
          "flex": 1,
          "fontSize": 15,
          "height": 40,
          "paddingHorizontal": 20,
        },
        null,
      ]
    }
    value=""
  />
  <View
    style={
      Object {
        "width": 40,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 15,
          "marginLeft": 10,
        }
      }
    >
      in
    </Text>
  </View>
</View>
`;
